### PR TITLE
[SPARK-21374][CORE] Fix reading globbed paths from S3 into DF with disabled FS cache

### PR DESCRIPTION
**Pull request by zsxwing**
## What changes were proposed in this pull request?

This PR replaces #18623 to do some clean up.

Closes #18623

## How was this patch tested?

Jenkins

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -638,4 +625,28 @@ object DataSource extends Logging {
     CatalogStorageFormat.empty.copy(
       locationUri = path.map(CatalogUtils.stringToURI), properties = optionsWithoutPath)
   }
+
+  /**
+   * If `path` is a file pattern, return all the files that match it. Otherwise, return itself.
+   * If `checkFilesExist` is `true`, also check the file existence.
+   */
+  private def checkAndGlobPathIfNecessary(
```
Comment (zsxwing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala]: Created a new method to make the codes cleaner.

**Review by gatorsmile — APPROVED** [review]

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Comment by gatorsmile** [comment]
Thanks! Merging to master.

**Comment by zsxwing** [comment]
Also merged to branch-2.2.

**Pull request closed without merging**